### PR TITLE
Add more detail on how to write adapters.

### PR DESCRIPTION
**Pull request by mstemm**
Take most of the general discussion from the gmail adapter docs at
https://github.com/juttle/juttle-gmail-adapter/blob/master/docs/adapter_impl_notes.md
and add them to this doc. Specifically, added the following:

 - How juttle loads an adapter
 - General discussion of live/historical and how that relates to the data read/written by an adapter.
 - Notes on configuration.
 - Options supported by many read commands.
 - Notes on the filter expression and how its parsed version is made available to the adapter.
 - Notes on write and expectations for write (don't block event loop) and eof (return a promise that resolves when everything is written).
 - Logging
 - Returning errors.

Once this is merged I'll change the gmail adapter docs to take out the
general stuff and tailor it more closely to the specific
choices/options used by the gmail adapter.

@demmer @dmehra @rlgomes

### CODE REVIEW

**Comment by mstemm** [comment]
@davidvgalbraith

File: docs/adapters/adapter_api.md [hunk 1]
```diff
@@ -6,9 +6,65 @@ NOTE: This API is currently under active development and is highly likely to cha
 
 __TOC__
 
+## Overview
+
+An Adapter is the interface between the juttle runtime and a backend data source. The backend consists of a set of data points. Each data point contains a number of fields. In most cases, each data point also contains a timestamp.
```
Comment (davidvgalbraith) [inline docs/adapters/adapter_api.md]: "The backend consists of a set of data points" -- I don't know what this means, thought a backend was like Elasticsearch or SQL?

File: docs/adapters/adapter_api.md [hunk 2]
```diff
@@ -6,9 +6,65 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+
+For reads, the adapter's job is to interpret the options included in the juttle `read` command into a set of matching data points, construct json objects representing those data points, and pass them to the juttle runtime. For writes, the adapter's job is to take the output of programs and "save" that output by writing data points to the backend.
```
Comment (davidvgalbraith) [inline docs/adapters/adapter_api.md]: Think it's Javascript objects, not "json objects".

File: docs/adapters/adapter_api.md [hunk 3]
```diff
@@ -6,9 +6,65 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+
+More sophisticated adapters work together with the juttle optimizer to push aggregation operations directly into the backend. For example, to count the number of data points in a given time period you could simply fetch all the data points and have the juttle program perform the counting. However, if supported by the backend it would be more efficient to count the number of data points directly in the backend and return the count instead.
```
Comment (davidvgalbraith) [inline docs/adapters/adapter_api.md]: Is there any in-depth description of the optimizer that we can link to here?

File: docs/adapters/adapter_api.md [hunk 4]
```diff
@@ -6,9 +6,65 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+
+## Writing an Adapter
+
+Here are the main steps involved in writing a new adapter:
+
+1. Create a javascript module that is require()d when a juttle program executes a `read <adapter>` or `write <adapter>`.
```
Comment (davidvgalbraith) [inline docs/adapters/adapter_api.md]: The "that is require()d when a juttle program executes a `read <adapter>` or `write <adapter>`." is pretty intense for a step 1 and it's not clear how to do that. How about just `Create a javascript module`?

File: docs/adapters/adapter_api.md [hunk 5]
```diff
@@ -6,9 +6,65 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+2. In that module, implement an init() function to define the read/write capabilities of the adapter.
```
Comment (davidvgalbraith) [inline docs/adapters/adapter_api.md]: I don't think an `init` function is required.

File: docs/adapters/adapter_api.md [hunk 6]
```diff
@@ -6,9 +6,65 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+3. Implement ES6 classes deriving from AdapterRead/AdapterWrite that perform the work of reading from/writing to the backend data source.
+4. The juttle runtime will instantiate objects using the classes to read from/write to the data source.
```
Comment (davidvgalbraith) [inline docs/adapters/adapter_api.md]: This doesn't involve any work on the implementers' part. It seems like steps 1, 2 and 4 can just go away and we should split up step 3 here.

File: docs/adapters/adapter_api.md [hunk 7]
```diff
@@ -6,9 +6,65 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+
+An adapter is packaged as a npm module. When the adapter is loaded, the CLI/juttle-engine perform a `require` of the module (i.e. the top-level directory). In [package.json](https://github.com/juttle/juttle-adapter-template/blob/master/package.json), the module defines a `main` of [lib/index.js](https://github.com/juttle/juttle-adapter-template/blob/master/lib/index.js), so `lib/index.js` is loaded.
+
+The main function exported by `lib/index.js` takes a `config` argument containing the configuration object for the adapter, and returns an object with `name`, `read`, `write`, and `optimize` attributes:
+
+```
+{
+    name: <adapter name>
+    read: <read adapter>,
+    write: <write adapter>,
```
Comment (davidvgalbraith) [inline docs/adapters/adapter_api.md]: How about <read module> and <write module> here?

**Comment by davidvgalbraith** [comment]
How does it look if we don't talk about optimization until after we've fully covered regular read? It seems like that'd give easier context for understanding.

File: docs/adapters/adapter_api.md (hunk 1 above)
Comment (mstemm) [inline docs/adapters/adapter_api.md]: How about contains a set of data points?

File: docs/adapters/adapter_api.md (hunk 3 above)
Comment (mstemm) [inline docs/adapters/adapter_api.md]: Not that I can see, I don't see any docs or anything in the wiki related to optimization.

File: docs/adapters/adapter_api.md (hunk 5 above)
Comment (mstemm) [inline docs/adapters/adapter_api.md]: It doesn't have to be init but it does have to be the export of the module, right? I can reword as "From that module, export an initialization function to define..."

File: docs/adapters/adapter_api.md (hunk 6 above)
Comment (mstemm) [inline docs/adapters/adapter_api.md]: I'll remove step 4, but I think it's still useful to talk about the module and its exported init function. I'll discuss with you in person.

File: docs/adapters/adapter_api.md (hunk 7 above)
Comment (mstemm) [inline docs/adapters/adapter_api.md]: How about <read class> and <write class>? They're being new() ed directly by the runtime.

**Comment by mstemm** [comment]
I think the current level of detail for optimization is ok--it's just talking about the idea of optimization in the overview and mentioning what the optimize attribute of the object returned from the init function should be. I think if we added more details on optimization, it should definitely come after read.

File: docs/adapters/adapter_api.md [hunk 8]
```diff
@@ -6,9 +6,64 @@ NOTE: This API is currently under active development and is highly likely to cha
 
 __TOC__
 
+## Overview
+
+An Adapter is the interface between the juttle runtime and a backend data source. The backend contains of a set of data points. Each data point contains a number of fields. In most cases, each data point also contains a timestamp.
```
Comment (dmehra) [inline docs/adapters/adapter_api.md]: Still oddly phrased, perhaps "The backend houses a data set, comprised of data points".

File: docs/adapters/adapter_api.md [hunk 9]
```diff
@@ -6,9 +6,64 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+
+For reads, the adapter's job is to interpret the options included in the juttle `read` command into a set of matching data points, construct Javascript objects representing those data points, and pass them to the juttle runtime. For writes, the adapter's job is to take the output of programs and "save" that output by writing data points to the backend.
```
Comment (dmehra) [inline docs/adapters/adapter_api.md]: Something's missing, maybe "to interpret the options included in the juttle `read` command to form a query to the backend, receive the matching data points, construct ..."

Comment (dmehra) [inline docs/adapters/adapter_api.md]: For writes, talking about "saving" the data doesn't cover all cases, if you "write http" that's hardly saving. How about "the adapter's job is to send the output of the juttle program to the backend by writing data points".

File: docs/adapters/adapter_api.md [hunk 10]
```diff
@@ -6,9 +6,64 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+
+More sophisticated adapters work together with the juttle optimizer to push aggregation operations directly into the backend. For example, to count the number of data points in a given time period you could simply fetch all the data points and have the juttle program perform the counting. However, if supported by the backend it would be more efficient to count the number of data points directly in the backend and return the count instead.
```
Comment (dmehra) [inline docs/adapters/adapter_api.md]: Let's add "That's exactly what many juttle adapters do."

File: docs/adapters/adapter_api.md [hunk 11]
```diff
@@ -6,9 +6,64 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+
+## Writing an Adapter
+
+Here are the main steps involved in writing a new adapter:
+
+1. Create a javascript module that can be require()d by the juttle runtime.
```
Comment (dmehra) [inline docs/adapters/adapter_api.md]: let's backtick `require()`d

File: docs/adapters/adapter_api.md [hunk 12]
```diff
@@ -6,9 +6,64 @@ NOTE: This API is currently under active development and is highly likely to cha
...
+```
+{
+    name: <adapter name>
+    read: <read adapter class>,
+    write: <write adapter class>,
+    optimize: <optimization module>
+}
+```
+
+The value for the `name` attribute corresponds to the `read <name>`/`write <name>` in juttle programs. The value for `read` is a class inheriting from [AdapterRead](../../lib/adapters/adapter-read.js) which performs the read work of the adapter. The value for `write` is an class inheriting from [AdapterWrite](../../lib/adapters/adapter-write.js) which performs the write work of the adapter. The value for `optimizer` is a module exporting functions `optimize_head`, `optimize_tail`, and `optimize_reduce` that do the work of optimizing the program.
```
Comment (dmehra) [inline docs/adapters/adapter_api.md]: why just those? there's at least `optimize_sort` aditionally

File: docs/adapters/adapter_api.md [hunk 13]
```diff
@@ -42,37 +97,61 @@ This object contains:
...
+read <adapter> [-from <moment>] [-to <moment>] [-timeField <field>] [-raw <expression>] [<filter expression>]
 ```
 
-
-## Read
+* `-from`/`-to`: [Juttle Moments](../../lib/runtime/types/juttle-moment.js) representing the start and end of the time range for the read.
+* `-last`: a JuttleMoment. shorthand for `from :now: - <last> -to :now:`
+* `-lag`: a JuttleMoment. Controls how long to wait behind real time to fetch datapoints. For example, with  `-from :1 minute ago: -to :now: -lag :30s:`, the runtime will initially wait 30 seconds, and then ask the points for the time period `[:1 minute ago:-:now:]`. A lag is useful when your backend takes a while to have results ready (due to write delays, etc).
+* `-timeField`: A field from the backend data that should be used as the time of the points emitted to the program.
+* `-raw`: A backend-specific search parameter that is passed opaquely to the adapter. For the Gmail Adapter, the `-raw` expression is passed directly through as a Gmail advanced search string.
```
Comment (dmehra) [inline docs/adapters/adapter_api.md]: should this mention `-every`?

**Comment by dmehra** [comment]
Maybe i missed it, we talked about including pointers to the wikis,
https://github.com/juttle/juttle/wiki/Adapter-Test-Checklist
https://github.com/juttle/juttle/wiki/Adapter-Publishing-Checklist

File: docs/adapters/adapter_api.md (hunk 12 above)
Comment (mstemm) [inline docs/adapters/adapter_api.md]: I took all the ones listed in the template. I don't see optimize_sort in lib/compiler/optimize. Is there somewhere else I can find the authoritative set?

File: docs/adapters/adapter_api.md (hunk 13 above)
Comment (mstemm) [inline docs/adapters/adapter_api.md]: Sure, it's in commonOptions, so I'll add it.

**Comment by mstemm** [comment]
@dmehra please look again?

**Comment by dmehra** [comment]
+1

**Pull request merged**